awscli/autocomplete/autogen.py

```diff
@@ -0,0 +1,177 @@
+            for key in identifier_map['resourceIdentifier']:
+                reverse_mapping[key] = name
+        op_map = {}
+        for op_name in op_names:
+            op_model = service_model.operation_model(op_name)
+            input_shape = op_model.input_shape
+            if not input_shape:
+                continue
+            for member in input_shape.members:
+                member_name = self._find_matching_member_name(
+                    member, reverse_mapping)
+                if member_name is None:
+                    continue
+                resource_name = reverse_mapping[member_name]
+                op = op_map.setdefault(op_name, {})
+                param = op.setdefault(member, {})
+                param['completions'] = [
+                    {'parameters': {},
+                     'resourceName': resource_name,
+                     'resourceIdentifier': member}
+                ]
+        return op_map
+
+    def _find_matching_member_name(self, member, reverse_mapping):
+        # Try to find something in the reverse mapping that's close
+        # to the member name.  This is really inefficient but this is
+        # a build time step so we don't really care about perf.
+        if member in reverse_mapping:
+            return member
+        matcher = SequenceMatcher()
+        matcher.set_seq2(member)
+        for candidate in reverse_mapping:
+            matcher.set_seq1(candidate)
+            match_ratio = matcher.ratio()
+            if match_ratio > 0.9:
+                return candidate
+
+    def _inject_resource(self, all_resources, resource):
+        r = all_resources.setdefault(resource.resource_name, {})
+        r['operation'] = resource.operation
+        ident = r.setdefault('resourceIdentifier', {})
+        ident[resource.ident_name] = resource.jp_expr
+
+    def _resource_for_single_operation(self, op_name, service_model):
+        # This heuristic works if there's a List/Describe operation
+        # that has only one shape in the output that's of type list.
+        # It will then find that inspect that list member to find
+        # the best guess at which identifier maps to the resource.
+        # This is a best effort attempt and relies heavily on naming
+        # conventions.
+        op_model = service_model.operation_model(op_name)
+        output = op_model.output_shape
+        list_members = [member for member, shape in output.members.items()
+                        if shape.type_name == 'list']
+        if len(list_members) != 1:
+            LOG.debug("Operation does not have exactly one list member, "
+                      "skipping: %s (%s)", op_name, list_members)
+            return
+        resource_shape_name = list_members[0]
+        list_member = output.members[resource_shape_name].member
+        op_with_prefix_removed = self._remove_verb_prefix(op_name)
+        if list_member.type_name == 'structure':
+            candidates = list_member.members
```

Comment:
It might be good to break these if statements into separate internal methods for readability